feat: 숙소 상세 조회 개발

decten opened the pull request:
### 💡Motivation
- 숙소 상세 조회를 개발 했습니다(필터링&stock 포함)
- 숙소 쿠폰 더 보기 기능을 개발했습니다

### 📌Changes
<img width="440" alt="image" src="https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/52107658/3162a641-13dd-40cd-9959-39c920d4e86a">
<img width="308" alt="image" src="https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/52107658/2a91afc3-112b-41b0-8a5a-9c2482d920cf">


### 🫱🏻‍🫲🏻To Reviewers
- 지금은 CQRS 패턴 적용 못 했고 포인트 하고 하려는데, 다른 분이 해주셔도 됩니다..

closes #24 #51

decten committed:
#9 feature&bug: 숙소 상세 조회의 filterRoomsByDate가 잘못 돼서 지금 room이 다 걸러짐

decten committed:
#9 fix: 로직 아닌 쿼리 기본 값 문제였음

기본 마감 기한(endDate)이 저장된 stock의 stopDate 보다 길어서 room이 다 걸러진 것

decten committed:
Merge branch 'develop' into feature/detail-accommodation

# Conflicts:
#	src/main/java/com/backoffice/upjuyanolja/domain/accommodation/service/AccommodationService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/entity/Coupon.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/repository/CouponRoomRepository.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/service/CouponService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/dto/response/RoomResponse.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/entity/RoomStatus.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/entity/RoomStock.java

decten committed:
Merge branch 'develop' into feature/detail-accommodation

decten committed:
#51 fix: 충돌 에러 해결

decten committed:
#51 refactor: 동일한 기능을 하는 BiFunction을 하나로 만듦

decten committed:
#51 refactor: 필요 없는 기능 제거

decten committed:
#51 refactor: CouponRoom이 CouponIssuance로 네이밍 바뀜

decten committed:
#51 refactor: CouponRoom이 CouponIssuance로 네이밍 바뀜

decten committed:
#51 perf: 쿠폰 상세 조회의 '쿠폰 더 보기 응답' dto 완성

decten committed:
#24 perf: 상세 조회의 '룸에 해당 하는 쿠폰' 응답 dto 완성

decten committed:
#51 feat: 상세 조회의 '룸' 응답 dto 완성

decten committed:
#51 feat: 룸에 해당 하는 'RoomStock' 리스트 리턴

decten committed:
#51 other: RoomStock 에러 코드 추가

decten committed:
#24 refactor&feat: Enum 원상 복구 및 필요 메서드 추가

decten committed:
#24 #51 feat: 필터링 없는 조회 잘 됨

decten committed:
#51 fix: soldout 반대로 표시하고 있어서 수정

decten committed:
Merge branch 'develop' into feature/detail-accommodation

# Conflicts:
#	src/main/java/com/backoffice/upjuyanolja/domain/accommodation/service/AccommodationCommandService.java

decten committed:
#51 refactor: null 대신 optional

decten committed:
Merge branch 'develop' into feature/detail-accommodation

decten committed:
Merge branch 'develop' into feature/detail-accommodation

# Conflicts:
#	src/main/java/com/backoffice/upjuyanolja/domain/accommodation/service/AccommodationCommandService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/service/CouponService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/service/RoomService.java
#	src/main/java/com/backoffice/upjuyanolja/global/exception/ErrorCode.java

decten committed:
#51 fix&refactor: 변수명 및 CQRS 패턴에 맞춰 빌드 에러 해결

decten commented on src/main/java/com/backoffice/upjuyanolja/domain/room/service/RoomQueryService.java:
```diff
@@ -32,7 +36,7 @@ public List<RoomImage> saveRoomImages(List<RoomImage> requests) {
     @Override
     @Transactional(readOnly = true)
-    public Room getRoomById(long roomId) {
+    public Room findRoomById(long roomId) {
```
repo에서 get이 아닌 find를 쓰고 있기 때문에 서비스 메서드 네이밍도 get이 아닌 find로 바꿨습니다

JeongUijeong replied:
ㅠㅠ 레포지토리 네이밍과 통일 시킬 필요가 없지 않나요 레포지토리 입장에서는 Room을 찾지만 서비스에서는 찾은 Room을 가져오는 것이기 때문에 get으로 다시 수정해주셨으면 좋겠습니다.

decten replied:
레포의 메서드가 find와 get이 모두 있어서, 서비스에서 get을 쓰면 레포에서도 get을 쓰고 있다고 혼동을 준다고 생각합니다

JeongUijeong replied:
저는 메서드를 사용하는 입장 기준으로 네이밍을 하는 편인데 find 보다는 get이 좀 더 적절하다고 생각합니다! 
그렇지만 현아님이 JPA 레포지토리에서 get을 사용하는 걸로 혼동을 준다고 하신 부분 참고해서 앞으로 QueryService 에서는 find를 쓰도록 하겠습니다!
다만, CommandService에서는 좀 더 의미상 적절하다고 생각하는 get으로 사용하도록 하겠습니다~ CommandService에서는 QueryService에서 찾은 엔터티를 받아올 뿐이며 찾는 비즈니스 로직이 포함되지 않기 때문입니다!

decten commented on src/main/java/com/backoffice/upjuyanolja/domain/room/service/usecase/RoomCommandUseCase.java:
```diff
@@ -3,10 +3,16 @@
 
     RoomInfoResponse saveRoom(Accommodation accommodation, RoomRegisterRequest request);
+
+    List<RoomStock> getFilteredRoomStocksByDate(Room room, LocalDate startDate, LocalDate endDate);
```
CQRS를 급하게 적용하다 보니 Dto를 못 만들었습니다.. 
나중에 List<RoomStock>을 담을 dto나 RoomStock을 담을 dto를 만들면 좋을 것 같습니다

JeongUijeong commented on src/docs/asciidoc/index.adoc:
```diff
@@ -21,7 +21,7 @@
 == API Common Response
 
-[[overview-status-code]]
+[[overview-couponStatus-code]]
```
왜 couponStatus로 수정하셨을까요??

decten replied:
헉 전체 바꾸면서 같이 바뀌었나 봅니다!
감사합니다

JeongUijeong commented on src/main/java/com/backoffice/upjuyanolja/domain/accommodation/controller/AccommodationController.java:
```diff
@@ -80,8 +81,8 @@ public ResponseEntity<SuccessResponse<AccommodationPageResponse>> getAccommodati
         @PathVariable Long accommodationId,
-        @RequestParam LocalDate startDate,
-        @RequestParam LocalDate endDate
+        @RequestParam(defaultValue = "2023-12-20", required = false) LocalDate startDate,
+        @RequestParam(defaultValue = "2023-12-21", required = false) LocalDate endDate
```
날짜는 현재 날짜 기준으로 String 파싱하여 defaultValue로 설정하시는 게 좋을 것 같습니다!

decten replied:
넵 혹시  제가 드린 sql에서도 roomStock 날짜를 now로 변경 해주실 수 있을까요?

JeongUijeong replied:
아직 해당 sql을 수정하지 않아서 수정 시 그 부분도 변경하도록 하겠습니닷! 😊

JeongUijeong commented on src/main/java/com/backoffice/upjuyanolja/domain/accommodation/service/AccommodationCommandService.java:
```diff
@@ -89,7 +88,6 @@ accommodation, getLowestPrice(accommodation.getId()),
     @Override
-    @Transactional(readOnly = true)
     public AccommodationOwnershipResponse getAccommodationOwnership(long memberId) {
         Member member = memberGetService.getMemberById(memberId);
```
@Transactional(readOnly=true)를 삭제 하셨는데 이유가 있나요?

JeongUijeong replied:
제가 작성한 조회 서비스 메서드에 다시 추가해주시길 바랍니다!

decten replied:
서비스 전체에 transactional 어노테이션이 있어서 삭제 했는데, 중복으로 사용 하는 건가요??

chaewon12 replied:
서비스 전체에 달린 Transactional은 쓰기 제어인데 
읽기만 수행하는 경우 readOnly 설정 하시려고 걸어두신 것 같아요~!
저는 readOnly=true 지정 하는게 성능상도 좋다고 알고 있습니다...!

chaewon12 replied:
[@Transactional(readOnly = true)를 왜 붙여야 하나요](https://hungseong.tistory.com/74)
이글 참고 하시면 좋을 것 같아요!

decten replied:
아하 감사합니다!
필요한 부분에 다시 추가 하겠습니다, 감사합니다😀

JeongUijeong reviewed (changes requested):
수고하셨습니다! 😊 몇 가지 코멘트 남겼습니다! 확인 부탁 드려요~

decten committed:
#51 refactor: restdocs 오타 수정

decten committed:
#51 feat: @RequestParam 오늘 날짜를 기본 값으로 설정

chaewon12 commented on src/main/java/com/backoffice/upjuyanolja/domain/room/service/RoomCommandService.java:
```diff
@@ -70,9 +72,24 @@ public RoomInfoResponse saveRoom(Accommodation accommodation, RoomRegisterReques
+    public List<RoomStock> getFilteredRoomStocksByDate(
+        Room room, LocalDate startDate, LocalDate endDate
+    ) {
+        return roomQueryUseCase.findStockByRoom(room).stream()
+            .filter(
+                stock ->
+                    !(stock.getDate().isBefore(startDate)) &&
+                        !(stock.getDate().isBefore(endDate)) &&
```
1. 날짜 필터링을 db 쿼리보다 stream().filter()로 하는게 더 빠른가요? 궁금해서 여쭤봅니다!
2.  !(stock.getDate().isBefore(endDate))은 endDate를 이후 값을 의미하지 않나요? 요구사항은 endDate 이전 값 아닌가요?

decten replied:
@chaewon12 
1. db 쿼리는 roomStock에서 room으로 조인 후 날짜를 찾아야 해서 stream을 사용 했습니다
2. 마감 일자 이후에도 stock이 있는 경우에 나오도록 필터링 한 건데, 마감 일자 이전까지만 값이 나와야 할까요?? 그렇다면 마감 일자에는 재고가 없는 경우가 나오는데 마감 일자에도 재고가 있는 경우로 나오게 하려고 했습니다!

chaewon12 replied:
저는 식을 이렇게 이해 했는데요 
 
```
!(stock.getDate().isBefore(startDate)) -> stock.getDate()가 startDate 이후 
&&
!(stock.getDate().isBefore(endDate)) ->  stock.getDate()가 endDate 이후 
```

그러면 startDate 이후 + endDate 이후  이면 결국 startDate 이후 의 값이 모두 필터링 되는 건데 

이렇게 의도하신게 맞으신가요? 맞으시다면 문제 없습니다!

decten replied:
startDate와 endDate 모두 이후 날짜 재고만 가져오려고 했습니다!
시작 날짜보다 이전 재고는 걸러 내고 시작 날짜 이후 재고가 있는 경우 가져옴 & 종료 날짜 이전 재고는 걸러 내고 종료 날짜 이후로도 재고가 있는 경우 가져옴
이렇게 생각했슴다

decten replied:
@chaewon12 근데 지금 필터링은 roomStock 날짜가 이렇게 반영 된다고 제가 생각 한 거라 
예약 구현하시는 채원님께서 날짜 필터링을 어떻게 하면 좋을지 말씀 해주시면 말씀 해주신 대로 로직을 변경 하겠습니다😀

chaewon12 replied:
예약에서는 startDate~ endDate 사이의 재고를 모두 가져와야하고
endDate 이후는 필요가 없어서 아마도 제가 쓸수는 없을 것 같아요...!
count 필터링도 없어야해서... 

그리고 예약 로직에서 따로 제가 작성해둬서 예약 쪽 고려 안하시고 짜시면 될것 같습니당

decten replied:
아하 감사합니다! 로직 수정 하겠습니다ㅎㅎ
<img width="273" alt="image" src="https://github.com/Upjuyanolja/Upjuyanolja_BE/assets/52107658/31b80808-b51f-4a3f-9e7d-30d3bb4246eb">

chaewon12 reviewed (changes requested):
코드 잘 봤습니다 ~! 
stock endDate 필터링 관련 궁금증 해결 되면 approve 하도록 하겠습니다 !

JeongUijeong committed:
#83 feat: 숙소 주소 데이터에 우편번호 추가

JeongUijeong committed:
#83 refactor: 숙소 우편번호 추가에 따른 코드 수정

JeongUijeong committed:
#79 refactor: 예약 기능 테스트에 우편 번호 추가 적용

matrixpower1004 committed:
Init: .gitignore 파일에 Querydsl의 QClass 생성 폴더 제외 추가

matrixpower1004 committed:
#86 feat: 쿠폰 만들기(View Data) 기능 개발

- 쿠폰 만들기(View Data) 기능 개발 완료
- 쿠폰 관련 엔티티 수정
- 요청되는 숙소 식별자 검증을 위한 커스텀 애너테이션 추가
- ValidationException을 처리하도록 handler에 등록

matrixpower1004 committed:
#86 test: 쿠폰 만들기(View Data)  API 테스트

- 테스트 전 반복되는 테스트 데이터 만들기를 sql에서 insert로 처리하도록 구현.

matrixpower1004 committed:
#86 docs: 쿠폰 만들기(View Data) API 문서화

matrixpower1004 committed:
#86 refactor: 쿠폰 할인 관련 데이터를 application.yml에서 가져 오도록 수정

- 쿠폰 할인 관련 정책 데이터를 enum 하드 코딩 -> application.yml 파일에서 가져오도록 수정
- 쿠폰 할인 관련한 정책이 변경되더라도 앱을 재배포 하거나 rebuild 하지 않아도 되도록 구현.
- 엔티티 필드명 수정에 따른 코드 수정.

matrixpower1004 committed:
Updated submodule submodule-data

matrixpower1004 committed:
Update coupon-api.adoc

matrixpower1004 committed:
Merge pull request #87 from Upjuyanolja/feature/coupon-make

#86 feat: 쿠폰 만들기(View Data) API 구현 완료 外

decten committed:
#51 perf: @Transactional(readOnly = true) 추가

decten committed:
Merge branch 'develop' into feature/detail-accommodation

# Conflicts:
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/dto/response/CouponAccommodationResponse.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/dto/response/CouponDetailResponse.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/entity/DiscountType.java
#	src/main/java/com/backoffice/upjuyanolja/domain/coupon/service/CouponService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/service/RoomCommandService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/service/RoomQueryService.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/service/usecase/RoomCommandUseCase.java
#	src/main/java/com/backoffice/upjuyanolja/domain/room/service/usecase/RoomQueryUseCase.java
#	src/main/java/com/backoffice/upjuyanolja/global/exception/ErrorCode.java

decten commented:
@matrixpower1004 CouponDetailResponse에서 지상님 응답이랑 제 응답이랑 달라서 dto는 다른 거 새로 만들어서 쓰셔야 합니다
그리고 지상님은 backoffice 컨트롤러랑 서비스를 분리 하셨고, 팀장님은 통일해서 쓰시는데 도메인마다 다르게 가는 건지 궁금합니다 @JeongUijeong

decten committed:
#51 fix: 빌드 에러 해결

JeongUijeong committed:
Merge branch 'develop' into feature/accommodation-zipcode

JeongUijeong committed:
Merge pull request #84 from Upjuyanolja/feature/accommodation-zipcode

refactor: 숙소 데이터에 우편번호 추가

chaewon12 committed:
#91 fix: submodule copy 분리

chaewon12 committed:
Merge pull request #92 from Upjuyanolja/fix/submodule-copy

fix: submodule copy 분리

decten committed:
#51 fix: 빌드 에러 해결

decten committed:
#51 fix: 빌드 에러 해결-로컬과 develop 병합

decten committed:
#51 fix: 빌드 에러 해결-로컬과 develop 병합

JeongUijeong commented:
DTO가 많으면 분리하고 그게 아니라면 같이 쓰면 되는 부분이라고 생각합니다..!
-컨트롤러와 서비스 DTO 분리에 대해-

decten committed:
#51 fix: 빌드 에러 해결-백오피스 쿠폰 테스트 코드 삭제

decten committed:
#51 feat: 재고 필터링 수정

decten commented:
아 dto는 같은 쿠폰 response 파일을 필드 수정하셔서 사용하시는 것 같아 새로 파일을 작성해야 한다는 말이었슴다
근데 지금은 제가 사용하는 버전만 있어서 나중에 새로 응답 dto 작성 해주시면 될 것 같습니다

decten committed:
#51 fix: sql문 삭제

decten commented:
브랜치 포기 새로 다시 오픈 하겠습니다ㅠㅠ
